solr/core/src/java/org/apache/solr/servlet/RequestRateLimiter.java

```diff
@@ -58,18 +58,18 @@ public RequestRateLimiter(RateLimiterConfig rateLimiterConfig) {
   public Pair<Boolean, AcquiredSlotMetadata> handleRequest() throws InterruptedException {
 
     if (!rateLimiterConfig.isEnabled) {
-      return new Pair<Boolean, AcquiredSlotMetadata>(true, null);
+      return new Pair<Boolean, AcquiredSlotMetadata>(true, new AcquiredSlotMetadata(null, null));
     }
 
     if (guaranteedSlotsPool.tryAcquire(rateLimiterConfig.waitForSlotAcquisition, TimeUnit.MILLISECONDS)) {
-      return new Pair<Boolean, AcquiredSlotMetadata>(true, new AcquiredSlotMetadata(this, false));
+      return new Pair<Boolean, AcquiredSlotMetadata>(true, new AcquiredSlotMetadata(this, guaranteedSlotsPool));
     }
 
     if (borrowableSlotsPool.tryAcquire(rateLimiterConfig.waitForSlotAcquisition, TimeUnit.MILLISECONDS)) {
-      return new Pair<Boolean, AcquiredSlotMetadata>(true, new AcquiredSlotMetadata(this, true));
+      return new Pair<Boolean, AcquiredSlotMetadata>(true, new AcquiredSlotMetadata(this, borrowableSlotsPool));
     }
 
-    return new Pair<Boolean, AcquiredSlotMetadata>(false, null);
+    return new Pair<Boolean, AcquiredSlotMetadata>(false, new AcquiredSlotMetadata(null, null));
   }
```

Comment:
I would write this as something like:

```
class SemaphoreWrapper() {
  Semaphore wrapped;

  release() {
    if (wrapped != null) wrapped.release();
  }
}

  public SempahoreWrapper handleRequest() throws InterruptedException {
      if (!rateLimiterConfig.isEnabled) {
       return nopPool; // = new SemaphoreWrapper(null);
     }

    if (guaranteedSlotsPool.tryAcquire(rateLimiterConfig.waitForSlotAcquisition, TimeUnit.MILLISECONDS)) {
       return new SemaphoreWrapper(guaranteedSlotsPool);
     }

    if (borrowableSlotsPool.tryAcquire(rateLimiterConfig.waitForSlotAcquisition, TimeUnit.MILLISECONDS)) {
       return new SemaphoreWrapper(borrowableSlotsPool);
     }
 
      return null;
   }
```

And probably have the limiter own the wrappers and return the right thing each time instead of creating a new wrapper.

Reply:
We need the rate limiter to be a part of the returned value here so that the decrement can be invoked on the right RequestRateLimiter?